### PR TITLE
Refactor Currency Settings

### PR DESCRIPTION
**Pull request by swansontec**
This is in preparation for the new Bitcoin plugin.

@samholmes, what is the format for the new nodes settings in your plugin? I just guessed when I wrote the `asCustomNodes` cleaner (see the last commit).

@thehobbit85 You can see where I have some commented out lines for "foo" and "bar" settings, which would also use the same structure as the existing custom nodes components.

#### PR Requirements

If you have made **any** visual changes to the GUI. Make sure you have:
- [x] Tested on iOS Tablet
- [ ] Tested on small Android
- [ ] n/a

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1202278972928633

### CODE REVIEW

File: src/components/themed/MaybeCustomNodesSetting.js [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    fontFamily: theme.fontFaceDefault,
+    fontSize: theme.rem(1),
+    paddingHorizontal: theme.rem(0.5),
+    textAlign: 'left'
+  }
+}))
+
+const asCustomNodesSetting: Cleaner<CustomNodesSetting> = asObject({
+  enableCustomNodes: asOptional(asBoolean, false),
+  customNodes: asArray(asString)
```
Comment (samholmes) [inline src/components/themed/MaybeCustomNodesSetting.js:107]: In `edge-currency-plugins`, the name of the field is `serverList`: https://github.com/EdgeApp/edge-currency-plugins/blob/master/src/common/plugin/pluginState.ts#L23

**Review by samholmes — APPROVED** [review]
Approved with optional changes. Also the relevant piece in edge-currency-plugins is located https://github.com/EdgeApp/edge-currency-plugins/blob/master/src/common/plugin/pluginState.ts

File: src/components/themed/MaybeCurrencySetting.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+|}
+
+type Props = {|
+  currencyConfig: EdgeCurrencyConfig
+|}
+
+/**
+ * Renders a settings section, but only if the cleaner passes.
+ */
+export function maybeCurrencySetting<T>(
```
[the remark this answers is not in the document]
Comment (swansontec) [reply src/components/themed/MaybeCurrencySetting.js:26]: The task was to pave the way for arbitrary settings, such as API keys, polling frequencies, or who knows what. All those sections would just use the HOC to slot into the scene in the same way. I don't know where to put HOC's, but we could agree on a location, I would put this in there.

File: src/components/themed/MaybeCustomNodesSetting.js (hunk 1 above)
Comment (swansontec) [reply src/components/themed/MaybeCustomNodesSetting.js:107]: OK, so your plugin needs to be using a cleaner. Also, we talked about changing the name so the upgrade goes smoothly. We don't want the new plugin trying to talk to Electrum servers leftover from the old plugin.

I am open to changing the shape of the cleaner, maybe naming the fields `customServers` and `enableCustomServers` if you prefer, but we want to agree before merging.

File: src/components/themed/MaybeCurrencySetting.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (swansontec) [reply src/components/themed/MaybeCurrencySetting.js]: Conclusion: `src/components/hoc`

File: src/components/themed/MaybeCustomNodesSetting.js (hunk 1 above)
Comment (swansontec) [reply src/components/themed/MaybeCustomNodesSetting.js]: https://github.com/EdgeApp/edge-currency-plugins/pull/311

Comment (samholmes) [reply src/components/themed/MaybeCustomNodesSetting.js]: I concur with the movement to change to your proposal. Very clean and clear semantics! 👍

File: src/components/hoc/MaybeCurrencySetting.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+type Props = {|
+  currencyConfig: EdgeCurrencyConfig
+|}
+
+/**
+ * Renders a settings section, but only if the cleaner passes.
+ */
+export function maybeCurrencySetting<T, X>(
+  Component: React$StatelessFunctionalComponent<CurrencySettingProps<T, X>>,
```
Comment (samholmes) [inline src/components/hoc/MaybeCurrencySetting.js:36]: Where does this type come from btw `React$StatelessFunctionalComponent`?

**Review by samholmes — APPROVED** [review]

File: src/components/hoc/MaybeCurrencySetting.js (hunk 3 above)
Comment (swansontec) [reply src/components/hoc/MaybeCurrencySetting.js:36]: It is built into Flow. It's a global, always-on symbol, just like `Array` or `$Call` are built in.

**Pull request merged**